flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -65,33 +71,40 @@ public void open(
                 this.checkpointedState == null,
                 "The " + getClass().getSimpleName() + " has already been initialized.");
 
-        this.checkpointedState =
-                context.getOperatorStateStore()
-                        .getListState(
-                                new ListStateDescriptor<>(
-                                        name + "-sequence-state", LongSerializer.INSTANCE));
-        this.valuesToEmit = new ArrayDeque<>();
-        if (context.isRestored()) {
-            // upon restoring
+        ListStateDescriptor<InternalState> stateDescriptor =
+                new ListStateDescriptor<>(
+                        name + "-sequence-state", TypeInformation.of(InternalState.class));
+        this.checkpointedState = context.getOperatorStateStore().getListState(stateDescriptor);
+        this.internalStates = Lists.newArrayList();
 
-            for (Long v : this.checkpointedState.get()) {
-                this.valuesToEmit.add(v);
-            }
+        if (context.isRestored()) {
+            checkpointedState.get().forEach(state -> internalStates.add(state));
         } else {
             // the first time the job is executed
-            final int stepSize = runtimeContext.getNumberOfParallelSubtasks();
             final int taskIdx = runtimeContext.getIndexOfThisSubtask();
-            final long congruence = start + taskIdx;
-
-            long totalNoOfElements = Math.abs(end - start + 1);
-            final int baseSize = safeDivide(totalNoOfElements, stepSize);
-            final int toCollect =
-                    (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+            final long stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            InternalState state = new InternalState(taskIdx, stepSize);
+            internalStates.add(state);
+        }
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
+        if (iterator.hasNext()) {
+            InternalState state = iterator.next();
+            long nextSequence =
+                    state.currentValue == 0 ? this.start + state.taskId : state.currentValue;
+            state.currentValue = nextSequence + state.stepSize;
+            // All sequence values are cleared from the stateList after they have been sent
+            if (state.currentValue > this.end) {
```

Comment:
If this.end is close to `Long.MAX_VALUE`, there's a possibility of overflow here with `nextSequence + state.stepSize`.

```suggestion
            if (state.currentValue > this.end || state.currentValue <= nextSequence) {
```

(I think this is correct logic, but `Math.addExact` could be used as well.)